kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/dek/AtomicCryptorUsageCounter.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filter.encryption.dek;
+
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.IntUnaryOperator;
+import java.util.function.LongUnaryOperator;
+
+/**
+ * <p>This counter exists to track different usages of a {@link Dek} so that we can
+ * destroy the underlying Key Material as soon as there are no cryptors using it.
+ * We identify two distinct usage types of a {@link Dek}, encryption and decryption.
+ * This classes responsibilities are:
+ * <ul>
+ *     <li>acquiring ( or denying ) new usages for each usage type</li>
+ *     <li>tracking outstanding usages for each usage type as usages are acquired and later released</li>
+ *     <li>tracking whether the application has destroyed each usage type</li>
+ *     <li>reporting when the end state has been reached (both usage types destroyed and with 0 usages outstanding)</li>
+ * </ul>
+ * </p>
+ * <p>Usages of a type can be acquired as long as that type is not destroyed.</p>
+ * <p>We use an AtomicLong, but it's really an atomic pair of ints acting as
+ * reference counts for the number of outstanding encryptors and decryptors.
+ * The counts start from 1 (i.e. 1 means no outstanding cryptor), increments with
+ * each cryptor created and decrements with each cryptor closed.
+ * Calling destroyFor(En|De)crypt flips the sign of the count
+ * and inverts the direction of the counting (i.e. cryptor close now increments).
+ * This means that we hit -1 when all outstanding cryptors have been closed.</p>
+ *
+ * <p>Here's a worked example:</p>
+ * <table>
+ *     <tr><th>Action</th>                          <th>Encryptor Count</th> <th>Decryptor Count</th></tr>
+ *     <tr><td>«{@link #START}»</td>                <td>1</td>               <td>1</td>    </tr>
+ *     <tr><td>{@link #acquireEncryptorUsage()}</td> <td>2</td>               <td>1</td>    </tr>
+ *     <tr><td>{@link #acquireEncryptorUsage()}</td> <td>3</td>               <td>1</td>    </tr>
+ *     <tr><td>{@link #acquireDecryptorUsage()}</td> <td>3</td>               <td>2</td>    </tr>
+ *     <tr><td>{@link #releaseEncryptorUsage()}</td><td>2</td>               <td>2</td>    </tr>
+ *     <tr><td>{@link #destroyForEncrypt()}</td>    <td>-2</td>              <td>2</td>    </tr>
+ *     <tr><td>{@link #releaseEncryptorUsage()}</td><td>-1</td>              <td>2</td>    </tr>
+ *     <tr><td>{@link #destroyForDecrypt()}</td>    <td>-1</td>              <td>-2</td>   </tr>
+ *     <tr><td>{@link #releaseDecryptorUsage()}</td><td>-1</td>              <td>-1</td>    </tr>
+ *     <tr><td colspan="3">«{@link #END}» // key gets destroyed </td></tr>
+ * </table>
+ */
+public class AtomicCryptorUsageCounter {
```

Comment:
I think atomic as an implementation detail. Its just a `CryptorUsageCounter`. Calling atomic might be salient if it was an interface and we had alternative implementations...

Reply:
Actually I think the atomic is about the most important thing about this class. It's part of how `Dek` is able to be thread-safe. But since it only has one use site it doesn't really matter. 

But certainly we should make this class package private.